python/ray/data/datasource/lancedb_datasource.py

```diff
@@ -0,0 +1,87 @@
+"""
+Module to read an LanceDB table into a Ray Dataset, by using the Ray Datasource API.
+"""
+
+import logging
+from typing import List, Optional, Union
+
+import lance
+import pyarrow as pa
+from lance import LanceFragment
+
+from ray.data import ReadTask
+from ray.data.block import Block, BlockMetadata
+from ray.data.datasource import Datasource
+from ray.util.annotations import DeveloperAPI
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class LanceDBDatasource(Datasource):
+    """LanceDB Datasource
+    Read a LanceDB table as a Ray Dataset
+
+    Parameters
+    ----------
+    uri : str
+        The base URI of the LanceDB dataset.
+    columns: list
+        A list of columns to return from the dataset.
+    filter: str
+        A standard SQL expressions as predicates for dataset filtering.
+    """
+
+    def __init__(
+        self,
+        uri: str,
+        columns: Union[list, None] = None,
+        filter: Union[str, None] = None,
+    ):
+        self.uri = uri
+        self.columns = columns
+        self.filter = filter
+
+        self.lance_ds = lance.dataset(uri)
+        self.fragments = self.lance_ds.get_fragments()
+
+    def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
+        # To begin with, read one Fragment at a time
+        # Each Ray Data Block contains a Pandas RecordBatch
+        def _read_single_fragment(fragment: LanceFragment) -> Block:
+            # Fetch batches from the fragment
+            batches = fragment.to_batches(columns=self.columns, filter=self.filter)
+
+            # Convert the generator of RecordBatch objects to a list
+            batches_list = list(batches)
+
+            # Convert the list of RecordBatch objects to a Table
+            table = pa.Table.from_batches(batches_list)
+
+            return table
+
+        read_tasks = []
+        for fragment in self.fragments:
+            data_files = ", ".join(
+                [data_file.path() for data_file in fragment.data_files()]
+            )
+
+            metadata = BlockMetadata(
+                num_rows=fragment.count_rows(),
+                size_bytes=None,
+                schema=fragment.schema,
+                input_files=[data_files],
+                exec_stats=None,
+            )
+
+            read_task = ReadTask(
+                lambda fragment=fragment: [_read_single_fragment(fragment)],
+                metadata,
+            )
```

Comment:
Fragments could be 100 GB or more. Maybe something we can implement later, but IMO it would be nice to let the user set the block size they want (in terms of # of rows), and then just slice the files according to that block size. We support partial scans of files.

Reply:
Agreed, "per-fragment" is probably too course for the long term.  Right now though it might be tricky to know, up front, exactly how many batches will be generated (will depend on the row group size of the fragment which I'm not sure we make available)

I think we could probably add an API that, given a read size, will tell you how many batches will be generated for a fragment.  Would that work?

Reply:
Alternatively, if ray supported the idea of a streaming source (e.g. a `ReadTask` that returns a `RecordBatchReader`) then a per-fragment API might actually work quite well.

Reply:
Yes, Ray Data supports streaming read in batches. Should we use https://lancedb.github.io/lance/read_and_write.html#iterative-read for per-fragment API?